cpp/src/arrow/dataset/dataset_internal.h

```diff
@@ -0,0 +1,207 @@
+        from_(std::move(from)),
+        to_(std::move(to)),
+        missing_columns_(to_->num_fields(), nullptr),
+        column_indices_(to_->num_fields(), kNoMatch),
+        scalars_(std::move(scalars)) {
+    DCHECK_EQ(scalars_.size(), missing_columns_.size());
+  }
+
+  Status Init() {
+    for (int i = 0; i < to_->num_fields(); ++i) {
+      const auto& field = to_->field(i);
+      int matching_index = from_->GetFieldIndex(field->name());
+      if (matching_index != kNoMatch) {
+        if (!from_->field(matching_index)->Equals(field)) {
+          return Status::Invalid("fields had matching names but were not equivalent ",
+                                 from_->field(matching_index)->ToString(), " vs ",
+                                 field->ToString());
+        }
+      } else {
+        auto out_type = to_->field(i)->type();
+        RETURN_NOT_OK(MakeArrayOfNull(pool_, out_type, 0, &missing_columns_[i]));
+      }
+
+      column_indices_[i] = matching_index;
+    }
+    return Status::OK();
+  }
+
+  Status Project(const RecordBatch& batch, std::shared_ptr<RecordBatch>* out) {
+    std::vector<std::shared_ptr<Array>> columns(to_->num_fields());
+
+    if (missing_columns_length_ < batch.num_rows()) {
+      RETURN_NOT_OK(ResizeMissingColumns(batch.num_rows()));
+    }
+
+    for (int i = 0; i < to_->num_fields(); ++i) {
+      int matching_index = column_indices_[i];
+      if (matching_index != kNoMatch) {
+        columns[i] = batch.column(matching_index);
+        continue;
+      }
+
+      columns[i] = missing_columns_[i]->Slice(0, batch.num_rows());
+    }
+
+    *out = RecordBatch::Make(to_, batch.num_rows(), std::move(columns));
+    return Status::OK();
+  }
+
+  const std::shared_ptr<Schema>& from() const { return from_; }
+
+  const std::shared_ptr<Schema>& to() const { return to_; }
+
+ private:
+  static constexpr int kNoMatch = -1;
+
+  Status ResizeMissingColumns(int64_t new_length) {
+    // TODO(bkietz) MakeArrayOfNull could use fewer buffers by reusing a single zeroed
+    // buffer for every buffer in every column which is null
+    for (int i = 0; i < to_->num_fields(); ++i) {
+      if (missing_columns_[i] == nullptr) {
+        continue;
+      }
+      if (scalars_[i] == nullptr) {
+        RETURN_NOT_OK(MakeArrayOfNull(pool_, missing_columns_[i]->type(), new_length,
+                                      &missing_columns_[i]));
+        continue;
+      }
+      RETURN_NOT_OK(
+          MakeArrayFromScalar(pool_, *scalars_[i], new_length, &missing_columns_[i]));
+    }
+    missing_columns_length_ = new_length;
+    return Status::OK();
+  }
+
+  MemoryPool* pool_;
+  std::shared_ptr<Schema> from_, to_;
+  int64_t missing_columns_length_ = 0;
+  std::vector<std::shared_ptr<Array>> missing_columns_;
+  std::vector<int> column_indices_;
+  std::vector<std::shared_ptr<Scalar>> scalars_;
+};
+
+constexpr int RecordBatchProjector::kNoMatch;
+
+class ProjectedRecordBatchReader : public RecordBatchReader {
```

Comment:
Add a comment that the Projector is dynamic in the sense that it adapts to the next record batch schema.